usr/src/uts/common/fs/zfs/dmu_send.c

```diff
@@ -2641,9 +3234,10 @@ receive_read_record(struct receive_arg *ra)
 	case DRR_OBJECT:
 	{
 		struct drr_object *drro = &ra->rrd->header.drr_u.drr_object;
-		uint32_t size = P2ROUNDUP(drro->drr_bonuslen, 8);
+		uint32_t size = DRR_OBJECT_PAYLOAD_SIZE(drro);
 		void *buf = kmem_zalloc(size, KM_SLEEP);
```

Comment:
I still need to dig into this more, but attempting to reproduce the panic seen by GernotS I was doing `zfs send unencrypted_zvol | ssh zfs recv encrypted_zvol` on a DEBUG kernel and got a warning:
```
WARNING: kmem_alloc(): sleeping allocation with size of 0; see kmem_zerosized_log for details
```
which seems to have been triggered by this alloc (based on the stack trace).  It has _not_ caused a panic (at least yet).  I plan to dig into this more to understand what's going on, but in case this is expected/acceptable behavior (i.e. red herring), or happens to jog the memory of someone more familiar with this code, I wanted to mention it.

Reply:
Thanks for working on reproducing the panic!

I believe that this warning can be safely ignored, though it would be good to address it anyway.

Reply:
Well I've managed to create _a_ panic :). I'm not sure yet if it's related to the crypto bits, but I was trying to `zfs send unencrypted | zfs send encrypted` between two 80Gb zvols and hit this:
```
panic[cpu2]/thread=fffffe001f3cdc20: assertion failed: aggsum_compare(&arc_meta_used, space) >= 0, file: ../../common/fs/zfs/arc.c, line: 2937
fffffe001f3cd970 genunix:process_type+16c03c ()
fffffe001f3cd9b0 zfs:arc_space_return+1b8 ()
fffffe001f3cda20 zfs:arc_free_data_impl+13d ()
fffffe001f3cda80 zfs:arc_free_data_buf+4e ()
fffffe001f3cdac0 zfs:arc_buf_destroy_impl+a9 ()
fffffe001f3cdb00 zfs:arc_buf_destroy+128 ()
fffffe001f3cdb70 zfs:dbuf_destroy+67 ()
fffffe001f3cdb90 zfs:dbuf_evict_one+f2 ()
fffffe001f3cdc00 zfs:dbuf_evict_thread+1af ()
fffffe001f3cdc10 unix:thread_start+8 ()
```
Of course the dump device was too small, so it didn't get saved :)

Reply:
(I've increased the size of the dump device and have restarted the same test that caused the panic, so hopefully I'll get something more useful).

Reply:
I've managed to recreate it again (this time with a saved crash dump).  This time the send/recv succeeded, and after doing a zfs destroy on the destination to start again, it paniced shortly after that.  It wasn't the same assertion, but was similar:
```
panic message: assertion failed: aggsum_compare(&arc_size, space) >= 0, file: ../../common/fs/zfs/arc.c, line: 2948
dump content: kernel pages only
> arc_size::print aggsum_t
{
    as_lock = {
        _opaque = [ 0 ]
    }
    as_lower_bound = 0xfffffffffffd9948
    as_upper_bound = 0x1948
    as_numbuckets = 0x8
    as_buckets = 0xfffffe170520ee00
}
> 0xfffffffffffd9948=e
                -157368
> ::stack
vpanic()
0xfffffffffbe15604()
arc_space_return+0x180(2000, 0)
```
I've not looked at this code much previously, so I'm not sure what this means yet (maybe something's forgetting to update stats leading to things being off?).  I can provide the dump to anyone interested (it's just a test VM setup to try to investigate some of these issues), as well as run any Dtrace, etc. as I continue my own digging.

Reply:
I was getting panics just like this (but without send/recv) on my freebsd port; mmacy went over it, and found some ZoL changes I'd missed, and that seemed to fix them up.  However, now we're not sure which specific changes those were, so he's looking over them again.

Reply:
I cherry-picked from this PR (the 3 commits) onto illumos-joyent (which syncs w/ illumos-gate daily) -- I also didn't see any differences that stood out.  I haven't looked at the ZoL tree though.  If there are some ZoL commits that are aren't in openzfs that this depends on, I'm sure we'll want to get those identified so they can get added as well.

Reply:
I've still had no success yet in tracking down what's leading to tripping the aggrsum asserts.  I have noticed in the dumps a similar pattern:

```
vpanic()
0xfffffffffbe15604()
arc_space_return+0x180(2000, 0)
arc_free_data_impl+0xc5(fffffe1d3ea2c448, 2000, fffffe1d3ea2c448)
arc_free_data_abd+0x2a(fffffe1d3ea2c448, fffffe17caad07f0, 2000, fffffe1d3ea2c448)
arc_hdr_free_pabd+0xa2(fffffe1d3ea2c448, 1)
arc_hdr_destroy+0x203(fffffe1d3ea2c448)
arc_freed+0x93(fffffe171b62b000, fffffe46691dd080)
zio_free_sync+0xad(fffffe1764204d40, fffffe171b62b000, 7c00, fffffe46691dd080, 0)
dsl_scan_free_block_cb+0x80(fffffe1713307e40, fffffe46691dd080, fffffe170ac09f00)
bptree_visit_cb+0x64(fffffe171b62b000, 0, fffffe46691dd080, fffffe00211b7cb0, fffffe5091e9d200, fffffe00211b8910)
traverse_visitbp+0x4f1(fffffe00211b8680, fffffe5091e9d200, fffffe46691dd080, fffffe00211b7cb0)
traverse_visitbp+0x37f(fffffe00211b8680, fffffe5091e9d200, fffffe175c73cc80, fffffe00211b7d90)
traverse_visitbp+0x37f(fffffe00211b8680, fffffe5091e9d200, fffffe1b874ef200, fffffe00211b7e70)
traverse_visitbp+0x37f(fffffe00211b8680, fffffe5091e9d200, fffffe5091e9d240, fffffe00211b7f10)
traverse_dnode+0xaf(fffffe00211b8680, fffffe5091e9d200, ffffffffffffffff, 1)
traverse_visitbp+0x6d1(fffffe00211b8680, fffffe5091e8c800, fffffe4f7a1e9000, fffffe00211b80c0)
traverse_visitbp+0x37f(fffffe00211b8680, fffffe5091e8c800, fffffe198bf49000, fffffe00211b81a0)
traverse_visitbp+0x37f(fffffe00211b8680, fffffe5091e8c800, fffffe4f92ce7000, fffffe00211b8280)
traverse_visitbp+0x37f(fffffe00211b8680, fffffe5091e8c800, fffffe1d42566000, fffffe00211b8360)
traverse_visitbp+0x37f(fffffe00211b8680, fffffe5091e8c800, fffffe4deaddf000, fffffe00211b8440)
traverse_visitbp+0x37f(fffffe00211b8680, fffffe5091e8c800, fffffe5091e8c840, fffffe00211b84e0)
traverse_dnode+0xaf(fffffe00211b8680, fffffe5091e8c800, ffffffffffffffff, 0)
traverse_visitbp+0x78f(fffffe00211b8680, 0, fffffe00211b8860, fffffe00211b8720)
```
What's interesting is that the last blkptr that leads to the assert always have the similar pattern:
```
> fffffe46691dd080::blkptr
DVA[0]=<0:13686c4000:2000>
salt=51dfe746b872cb8f iv=18fa53b49aa0e31f:32af9199
[L0 ZVOL] FLETCHER_4 OFF encrypted LE contiguous unique single
size=2000L/2000P birth=31593L/31593P fill=1
```
It's always a L0 ZVOL blkptr that trips the assert.  All the other blkptrs in the stack are L1 or higher, or if L0 the type is DNODE and not ZVOL. e.g.
```
> fffffe5091e9d240::blkptr
DVA[0]=<0:c40682000:2000>
DVA[1]=<0:14c065c000:2000>
[L3 ZVOL] FLETCHER_4 LZ4 indirect-MAC LE contiguous unique double
size=20000L/2000P birth=31594L/31594P fill=5242880
cksum=11596bab76b97df7:ac761e281b7b08e4:f2ce78d25b1d72b2:91265f08e45c67a6
> fffffe4f7a1e9000::blkptr
DVA[0]=<0:6c1ef6000:4000>
DVA[1]=<0:1e41ef8000:4000>
salt=51dfe746b872cb8f iv=6dc8f0b2511e8bd0:21ff08b2
[L0 DNODE] FLETCHER_4 OFF encrypted LE contiguous unique double
size=4000L/4000P birth=31594L/31594P fill=2
cksum=3f29a6e9422ce2c:ead4b8958a6dd2a9:cd8aacc59f040598:f608bd4cc6e5e8a2
```
I've also tried the same test on a DEBUG image that where the only difference is missing this PR, and I have not been able to trigger the asserts.